### ISSUE TITLE
[QUESTION/SUGGESTION] Alphanumerical natural sorting

### ISSUE THREAD

**Issue by Deathnetworks** [litedb-org/LiteDB#2577]
Trying to sort text that contains numbers and can't seem to find away that doesn't involve loading the entire db and using C# to do the sorting via a comparer

```
records = (from x in col
                               .Query()
                               .OrderByDescending(x => x.title)
                               .Skip(pageskip)
                               .Limit(limit ?? 50)
                               .ToEnumerable()
```
                               
`SELECT $ FROM book order by title ASC`

Example results I'm getting -

- "x - Chapter 1"
- "x - Chapter 10"
- "x - Chapter 100"
- "x - Chapter 101"
- "x - Chapter 102"
- "x - Chapter 103"

How I'd prefer it be sorted

- "x - Chapter 1"
- "x - Chapter 2"
- "x - Chapter 3"
- ....4-9
- "x - Chapter 10"
- ...11-99
- "x - Chapter 100"
- "x - Chapter 101"
- "x - Chapter 102"

Any suggestions on accomplishing this feat?

Using 5.0.8

**Comment by dethknite** [comment]
This can be accomplished by making your own custom System.Collections.IComparer (casting an obj to your desired type if necessary). Advanced implementations would check the obj to verify what it is prior -- ie. int, datetime (specific formats), custom object T, etc... and perhaps fallback to string.

Something like:
(https://learn.microsoft.com/en-us/dotnet/api/system.icomparable-1?view=net-9.0)

**Comment by Deathnetworks** [comment]
Yup, that's the solution that requires loading the db then using c# to sort... 
Was hoping for a way to do it directly via the db so it's not as heavy/slow.

But that's exactly what I ended up running with, while using filtering to reduce the dataset.

**Comment by JKamsker** [comment]
There's no natural-sort collation in LiteDB (string comparison goes through the collation's culture and `CompareOptions`), but you can sort on a computed key inside the database. For your title pattern:

```sql
SELECT $ FROM book
ORDER BY INT(SUBSTRING(title, INDEXOF(title, 'Chapter ') + 8))
```

This gives 1, 2, 3, 10, 20, 100, 101, 102. `LPAD(SUBSTRING(...), 6, '0')` works too if you want the key to stay a string.

For paging, index that same expression so `Skip`/`Limit` walk the index instead of sorting everything:

```csharp
const string chapterKey = "INT(SUBSTRING($.title, INDEXOF($.title, 'Chapter ') + 8))";
col.EnsureIndex("chapterNo", chapterKey);

var page = col.Query()
    .OrderBy(chapterKey)          // or OrderByDescending(chapterKey)
    .Skip(pageskip)
    .Limit(limit ?? 50)
    .ToList();                    // GetPlan(): FULL INDEX SCAN(chapterNo), no separate sort
```

The simplest and most robust option is to store the number in its own field (e.g. `Chapter`), index it, and use `OrderBy(x => x.Chapter)`. General natural sorting over arbitrary text isn't built in. I checked this on 5.0.21 and the 6.x prerelease; it's worth upgrading from 5.0.8 in any case.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
